src/steiner_tree.rs

```diff
@@ -113,3 +120,165 @@ fn _metric_closure_edges(
+        .iter()
+        .map(|edge| pairwise(edge.path.clone()))
+        .flatten()
+        .filter_map(|x| x.0.map(|a| [a, x.1]))
+        .collect();
+    let out_edges: HashSet<(usize, usize)> =
+        out_edge_list.iter().map(|x| (x[0], x[1])).collect();
+    let mut out_graph = graph.clone();
+    let out_nodes: HashSet<NodeIndex> = out_edge_list
+        .iter()
+        .map(|x| x.iter())
+        .flatten()
+        .copied()
+        .map(NodeIndex::new)
+        .collect();
+    for node in graph
+        .graph
+        .node_indices()
+        .filter(|node| !out_nodes.contains(node))
+    {
+        out_graph.graph.remove_node(node);
+        out_graph.node_removed = true;
+    }
+    for edge in graph.graph.edge_references().filter(|edge| {
+        let source = edge.source().index();
+        let target = edge.target().index();
+        !out_edges.contains(&(source, target))
+            && !out_edges.contains(&(target, source))
+    }) {
+        out_graph.graph.remove_edge(edge.id());
+    }
+    // Deduplicate potential duplicate edges
+    if graph.multigraph {
+        // Find all edges between nodes
+        let mut duplicate_map: HashMap<
+            [NodeIndex; 2],
+            Vec<(EdgeIndex, &PyObject)>,
+        > = HashMap::new();
+        for edge in graph.graph.edge_references() {
+            if duplicate_map.contains_key(&[edge.source(), edge.target()]) {
+                duplicate_map
+                    .get_mut(&[edge.source(), edge.target()])
+                    .unwrap()
+                    .push((edge.id(), edge.weight()));
+            } else if duplicate_map
+                .contains_key(&[edge.target(), edge.source()])
+            {
+                duplicate_map
+                    .get_mut(&[edge.target(), edge.source()])
+                    .unwrap()
+                    .push((edge.id(), edge.weight()));
+            } else {
+                duplicate_map.insert(
+                    [edge.source(), edge.target()],
+                    vec![(edge.id(), edge.weight())],
+                );
+            }
+        }
+        // For a node pair with > 1 edge find minimum edge and remove others
+        for edges_raw in duplicate_map.values().filter(|x| x.len() > 1) {
+            let mut edges: Vec<(EdgeIndex, f64)> =
+                Vec::with_capacity(edges_raw.len());
+            for edge in edges_raw {
+                let res = weight_fn.call1(py, (edge.1,))?;
+                let raw = res.to_object(py);
+                let weight = raw.extract(py)?;
+                edges.push((edge.0, weight));
+            }
+            edges.par_sort_unstable_by(|a, b| {
+                let weight_a = a.1;
+                let weight_b = b.1;
+                weight_a.partial_cmp(&weight_b).unwrap_or(Ordering::Less)
+            });
+            edges[1..].iter().for_each(|x| {
+                out_graph.graph.remove_edge(x.0);
+            });
+        }
+    }
```

Comment:
I think it's reasonable to move this into a separate function.

Reply:
Done in: https://github.com/Qiskit/retworkx/pull/392/commits/7db336d9f7a4200e419a984bd1e2bdefebf49173